src/AspNet.Security.OAuth.XiaoMi/XiaoMiAuthenticationHandler.cs

```diff
@@ -0,0 +1,199 @@
+            ["clientId"] = Options.ClientId
+        };
+
+        var address = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, parameters);
+
+        using var response = await Backchannel.GetAsync(address);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving user information.");
+        }
+
+        var json = await response.Content.ReadAsStringAsync(Context.RequestAborted);
+        using var payload = JsonDocument.Parse(json);
+        var rootElement = payload.RootElement;
+
+        if (!rootElement.TryGetProperty("data", out JsonElement dataElement))
+        {
+            var errorCode = rootElement.GetProperty("code").GetString()!;
+            throw new Exception($"An error (Code:{errorCode}) occurred while retrieving user information.");
+        }
+
+        identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, dataElement.GetString("unionId")!, ClaimValueTypes.String, Options.ClaimsIssuer));
+
+        var principal = new ClaimsPrincipal(identity);
+        var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, dataElement);
+        context.RunClaimActions();
+
+        await Events.CreatingTicket(context);
+        return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+    {
+        var tokenRequestParameters = new Dictionary<string, string?>()
+        {
+            ["client_id"] = Options.ClientId,
+            ["client_secret"] = Options.ClientSecret,
+            ["code"] = context.Code,
+            ["redirect_uri"] = context.RedirectUri,
+            ["grant_type"] = "authorization_code",
+        };
+
+        // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+        if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+        {
+            tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        var address = QueryHelpers.AddQueryString(Options.TokenEndpoint, tokenRequestParameters);
+
+        using var response = await Backchannel.GetAsync(address);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.ExchangeCodeErrorAsync(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+        }
+
+        var json = await response.Content.ReadAsStringAsync(Context.RequestAborted);
+        var payload = JsonDocument.Parse(json.Replace("&&&START&&&", string.Empty, StringComparison.OrdinalIgnoreCase));
```

Comment:
What's this `&&&START&&&` for?

Reply:
This is required by (the documentation)[https://dev.mi.com/console/doc/detail?pId=707], but I think it's bad.

Reply:
Wow 😶

Reply:
I agree that's also _not good_, however I think the best way to deal with it would be:

1. Do a `StartsWith()` and then trim it off if present. Then there's no danger (even if theoretical) of it replacing _other_ things in the payload that contain that string that shouldn't.
2. Add a comment explaining why the code works that way with a link to that documentation.

Reply:
I agree too, I will do it, thank you!

Reply:
Are we 100% sure it's not a typo in the docs? If you use Fiddler to intercept the request or put a breakpoint at line 99, do you see the `&&&START&&&` prefix in the returned JSON payload?

Reply:
@kevinchalet  Yes, I can be sure that the document is ok, when I debug I can see the prefix in the returned JSON payload.

![image](https://user-images.githubusercontent.com/35516482/167077636-47ec96c5-bf7d-44ea-a62e-b035bdabceb9.png)